go/plugins/googleai/googleai.go

```diff
@@ -16,17 +16,154 @@ package googleai
+}
+
+// Config configures the plugin.
+type Config struct {
+	// API key. Required.
+	APIKey string
+	// Generative models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Models []string
+	// Embedding models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Embedders []string
+}
+
+func Init(ctx context.Context, cfg Config) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("googleai.Init: %w", err)
+		}
+	}()
+
+	if err := initClient(ctx, cfg.APIKey); err != nil {
+		return err
+	}
+
+	needModels := len(cfg.Models) == 0
+	needEmbedders := len(cfg.Embedders) == 0
+	if needModels || needEmbedders {
+		iter := state.client.ListModels(ctx)
+		for {
+			mi, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			// Model names are of the form "model/name".
+			name := path.Base(mi.Name)
+			if needModels && slices.Contains(mi.SupportedGenerationMethods, "generateContent") {
+				cfg.Models = append(cfg.Models, name)
+			}
+			if needEmbedders && slices.Contains(mi.SupportedGenerationMethods, "embedContent") {
+				cfg.Embedders = append(cfg.Embedders, name)
+			}
+		}
+	}
+	state.models = map[string]ai.Generator{}
+	for _, name := range cfg.Models {
+		state.models[name] = defineModel(name, state.client)
+	}
+
+	state.embedders = map[string]ai.Embedder{}
+	for _, name := range cfg.Embedders {
+		state.embedders[name] = defineEmbedder(name, state.client)
+	}
+	return nil
+}
+
+func initClient(ctx context.Context, apiKey string) error {
+	if apiKey == "" {
+		return errors.New("missing API key")
+	}
+
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	if state.client != nil {
+		return errors.New("already initialized")
+	}
+	c, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
+	if err != nil {
+		return err
+	}
+	state.client = c
+	return nil
+}
+
+func defineModel(name string, client *genai.Client) ai.Generator {
+	meta := &ai.GeneratorMetadata{
+		Label: "Google AI - " + name,
+		Supports: ai.GeneratorCapabilities{
+			Multiturn: true,
+		},
+	}
+	g := generator{model: name, client: client}
+	return ai.DefineGenerator("google-genai", name, meta, g.Generate)
+}
+
+func defineEmbedder(name string, client *genai.Client) ai.Embedder {
+	return ai.DefineEmbedder("google-genai", name, func(ctx context.Context, input *ai.EmbedRequest) ([]float32, error) {
+		em := client.EmbeddingModel(name)
+		parts, err := convertParts(input.Document.Content)
+		if err != nil {
+			return nil, err
+		}
+		res, err := em.EmbedContent(ctx, parts...)
+		if err != nil {
+			return nil, err
+		}
+		return res.Embedding.Values, nil
+	})
+}
+
+// Generator returns the generator with the given name.
+// It panics if the generator was not configured.
+func Generator(name string) ai.Generator {
+	if g, ok := state.models[name]; ok {
+		return g
+	}
+	panic(fmt.Sprintf("googleai: generator %q not configured", name))
```

Comment:
Calling `panic` here seems pretty unfriendly.  Why not return `nil`?

Reply:
Done.